pkg/container/versioned/value.go

```diff
@@ -0,0 +1,440 @@
+			return errStaleVersion
+		}
+
+		if node.versions.contains(version) {
+			// link the new node after the current one
+			newNode.next.Store(node.next.Load())
+			node.next.Store(newNode)
+
+			// truncate the validity of this node to end at 'version' *after* the new
+			// node with validity starting from 'version' has been linked after it
+			// (above), so that either this or the new value is reachable at all times
+			// for readers with 'version'
+			node.versions.past.store(version)
+			break
+		}
+
+		node = node.next.Load()
+		prev = &node.next
+	}
+	if node == nil {
+		// Add the new value at the end
+		prev.Store(newNode)
+	}
+
+	return nil
+}
+
+// RemoveAt changes the validity of the stored value previously valid at version 'next' to have
+// ended at version 'next'.
+// 'next' must be later than any the current version visible to the readers.
+// Returns an error if this is not the case.
+// Callers must coordinate for mutual exclusion.
+func (v *Value[T]) RemoveAt(next NextVersion) error {
+	version := version(next)
+	if version == invalidVersion {
+		return errInvalidVersion
+	}
+
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			slog.Error("RemoveAt may not be called with version lower than existing already", logfields.NewVersion, version, logfields.OldVersion, node.versions.first, logfields.Stacktrace, stacktrace, logfields.LogSubsys, "containers/versioned")
+			return errStaleVersion
+		}
+
+		if node.versions.contains(version) {
+			// Truncate the validity of this node to end at 'version'.
+			// After this readers with 'version' and above no longer see this value,
+			// while readers with versions before 'version' still see this.
+			node.versions.past.store(version)
+			break
+		}
+	}
+	return nil
+}
+
+// RemoveBefore removes all values whose validity ends before 'keepVersion'.
+// Caller must coordinate for mutual exclusion.
+func (v *Value[T]) RemoveBefore(keepVersion KeepVersion) {
+	version := version(keepVersion)
+	// find all values that are no longer valid at 'version'
+	node := v.head.Load()
+	for node != nil && node.versions.past.load() <= version {
+		// This node is no longer visible for readers with 'version' and above,
+		// so this can be safely removed.
+		node = node.next.Load()
+	}
+	v.head.Store(node)
+}
+
+// At returns value of type 'T' valid for the given version, or an empty value if none is found.
+func (v *Value[T]) At(hold *VersionHold) T {
+	version := invalidVersion
+	if hold != nil {
+		version = hold.version
+	}
+	if version == invalidVersion {
+		stacktrace := hclog.Stacktrace()
+		slog.Error("At: Invalid VersionHold finds nothing", logfields.Stacktrace, stacktrace, logfields.LogSubsys, "containers/versioned")
+	}
+
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if node.versions.contains(version) {
+			return node.value
+		}
+	}
+	var empty T
```

Comment:
nit: slightly more elegant to name the return value `empty` so that you don't have to declare it 

```go
func (v *Value[T]) At(hold *VersionHold) (empty T) {
...
return
}
```

Reply:
Changed, but with `return empty` which IMO read better :-)